src/main/scala/com/fulcrumgenomics/bam/EstimatePoolingFractions.scala

```diff
@@ -0,0 +1,242 @@
+      new PoolingFractionMetric(
+        sample             = sample,
+        variant_sites      = sites,
+        singletons         = singletons,
+        estimated_fraction = fractions(index),
+        standard_error     = stderrs(index),
+        ci99_low           = max(0, fractions(index) - stderrs(index)*Ci99Width),
+        ci99_high          = min(1, fractions(index) + stderrs(index)*Ci99Width))
+    }
+
+    Metric.write(metrics, path=output)
+
+    if (regression.estimateRegressionParameters().exists(_ < 0)) {
+      logger.error("#################################################################################")
+      logger.error("# One or more samples is estimated to have fraction < 0. This is likely due to  #")
+      logger.error("# insufficient coverage at some loci and/or too few SNPs being used.            #")
+      logger.error("# Please try again with a higher min-coverage-per-sample.                       #")
+      logger.error("#################################################################################")
+      fail(1)
+    }
+  }
+
+  /** Verify a provided sample list, or if none provided retrieve the set of samples from the VCF. */
+  private def pickSamplesToUse(vcfReader: VCFFileReader): Array[String] = {
+    if (samples.nonEmpty) {
+      val samplesInVcf = vcfReader.getFileHeader.getSampleNamesInOrder.toSet
+      val missingSamples = samples.filterNot(samplesInVcf.contains)
+      if (missingSamples.nonEmpty) fail(s"Samples not present in VCF: ${missingSamples.mkString(", ")}")
+      else samples.toArray.sorted
+    }
+    else {
+      vcfReader.getFileHeader.getSampleNamesInOrder.toSeq.toArray.sorted // toSeq.toArray is necessary cos util.ArrayList.toArray() exists
+    }
+  }
+
+  /** Loads up and merges all the interval lists provided. Returns None if no intervals were specified. */
+  private def loadIntervals: Option[IntervalList] = {
+    if (this.intervals.isEmpty) {
+      None
+    }
+    else {
+      val list = IntervalList.fromFile(this.intervals.head.toFile)
+      val dict = list.getHeader.getSequenceDictionary
+      this.intervals.tail.foreach { f =>
+        val tmp = IntervalList.fromFile(f.toFile)
+        if (!SequenceUtil.areSequenceDictionariesEqual(dict, tmp.getHeader.getSequenceDictionary))
+          fail(s"Sequence dictionaries differ between ${this.intervals.head} and ${f}")
+        else
+          list.addall(tmp.getIntervals)
+      }
+
+      Some(list.uniqued(false))
+    }
+  }
+
+  /** Generates an iterator over non-filtered bi-allelic SNPs where all the required samples are genotyped. */
+  def constructVcfIterator(in: VCFFileReader, intervals: Option[IntervalList], samples: Array[String]): Iterator[VariantContext] = {
+    val vcfIterator = intervals match {
+      case None     => in.toIterator
+      case Some(is) => ByIntervalListVariantContextIterator(in, is)
```

Comment:
Just so you aware, there are two underlying implementations, and the one you are going to use skips around the vcf using the index, if that's ok.